lib/openhab/core/items/semantics.rb

```diff
@@ -390,6 +397,67 @@ def points(*point_or_property_types)
           result.delete(self)
           result
         end
+
+        # @deprecated OH3.4 - this check is only needed for OH3.4
+        if org.openhab.core.semantics.SemanticTags.respond_to?(:add)
+
+          #
+          # Adds custom semantic tags.
+          #
+          # @return [Semantics::Tag] The added semantic tag class
+          #
+          # @overload self.add(**tags)
+          #   Quickly add one or more semantic tags using the default label, empty synonyms and descriptions.
+          #
+          #   @param [kwargs] **tags One or more `tag` => `parent` pairs
+          #   @return [Boolean] true if all tags were added successfully
+          #
+          #   @example Add one semantic tag `Balcony` whose parent is `Semantics::Outdoor` (Location)
+          #     Semantics.add(Balcony: Semantics::Outdoor)
+          #
+          #   @example Add multiple semantic tags
+          #     Semantics.add(Balcony: Semantics::Outdoor,
+          #                   SecretRoom: Semantics::Room,
+          #                   Motion: Semantics::Property)
+          #
+          # @overload self.add(label: nil, synonyms: "", description: "", **tags)
+          #   Add a custom semantic tag with extra details.
+          #
+          #   @example
+          #     Semantics.add(SecretRoom: Semantics::Room, label: "My Secret Room",
+          #       synonyms: "HidingPlace", description: "A room that requires a special trick to enter")
+          #
+          #   @param [String,nil] label Optional label. When nil, infer the label from the tag name,
+          #     converting `CamelCase` to `Camel Case`
+          #   @param [String] synonyms A comma separated list of synonyms for this tag.
+          #   @param [String] description A longer description of the tag.
+          #   @param [kwargs] **tags Exactly one pair of `tag` => `parent`
+          #   @return [Boolean] true if the tag was added successfully
+          #
+          def self.add(label: nil, synonyms: "", description: "", **tags)
+            raise "Tags must be specified" if tags.empty?
+            if (tags.length > 1) && !(label.nil? && synonyms.empty? && description.empty?)
+              raise "Additional options can only be specified when creating one tag"
+            end
+
+            tags.map do |name, parent|
+              parent =  parent.is_a?(Symbol) ? parent.to_s : parent.java_class
+              name = name.to_s
+              org.openhab.core.semantics.SemanticTags.add(name, parent, label, synonyms, description)
+              const_missing(name)
+            end.any?(&:!)
+          end
+
+          #
+          # Automatically looks up new semantic classes and adds them as `constants`
+          #
+          # @return [Tag, nil]
+          #
+          def self.const_missing(sym)
```

Comment:
should maybe remove the old code that manually iterates and sets them, relying solely on this?

Reply:
Could do that, but we'll lose the ability to do `Semantics.constants.include?(:Room)`. WDYT?

Reply:
Yes, I think we should. Otherwise it works differently depending on if a tag was defined before or after a script initially loads. We could also override Semantics.constants to look them all up.

Reply:
How would it work differently?

Reply:
Because as is a tag added after a script loads won't be in Semantics.constants until you reference it

Reply:
The only way to look them up is by traversing through those streams. The core's SemanticTags.add() would insert the newly added tags into those streams so it would work fine even if some tags were added before the script loads.

I'm not sure how to override Semantics.constants - that sounds complicated

Reply:
> Because as is a tag added after a script loads won't be in Semantics.constants until you reference it

Hmm... yes they will be in Semantics.constants but only if they were added in the same scripts. When added by other scripts,.... no they won't be until you reference it.

I see, so we do need to override Semantics.constants. How do we do it efficiently? Should we simply do a reload every time?

Reply:
*after*, not before. If script A loads, without any custom tags, but with a rule that lists all constants in Semantics, then script B loads and defines a custom tag, then script A's rule triggers, it won't find the custom tag. If script A's rule instead directly references a tag, it will work.

Comparing to items which are Ruby constants, you can't look them up via the constants list, but only const_missing.

So I guess my vote is to _only_ use const_missing now, and then add a Semantics.tags method so you can list them all.

Reply:
> So I guess my vote is to _only_ use const_missing now, and then add a Semantics.tags method so you can list them all.

+1 This is a nice solution.

Reply:
I've just pushed Semantics.tags. I'm unfamiliar with the whole enumerable / enumerator. 
It seems that adding `.lazy` to java's Stream Spliterator converts it into an Enumerator::Lazy. Is that a usable object to be returned by `.tags` ?